[GTK] webrtc/datachannel/multiple-connections.html flaky crashes in RiceBackendProxy::create()

philn opened the pull request:
#### 1c1bd26fae15c77a3a4bf6e6a1f329df7844826c
<pre>
[GTK] webrtc/datachannel/multiple-connections.html flaky crashes in RiceBackendProxy::create()
<a href="https://bugs.webkit.org/show_bug.cgi?id=303366">https://bugs.webkit.org/show_bug.cgi?id=303366</a>

Reviewed by Xabier Rodriguez-Calvar.

Handle IPC issues in RiceBackendProxy::create() and propagate backend creation failure to the
IceAgent. This patch also fixes various leaks in the Rice backend, including GSource leaks.

Canonical link: <a href="https://commits.webkit.org/304159@main">https://commits.webkit.org/304159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d345c2069758464471e0254c46b4704a2733000a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134577 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7035 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142102 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86534 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136447 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7637 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6893 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102862 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70152 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137524 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5349 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120620 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83660 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/133926 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5206 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2821 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2704 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114440 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38761 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144797 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6710 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39343 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111261 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6784 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5630 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111543 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28329 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5037 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116895 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60585 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6761 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35088 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6573 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70344 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6808 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6684 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -671,7 +675,10 @@ WebKitGstIceAgent* webkitGstWebRTCCreateIceAgent(const String& name, ScriptExecu
 
     auto agent = reinterpret_cast<WebKitGstIceAgent*>(g_object_new(WEBKIT_TYPE_GST_WEBRTC_ICE_BACKEND, "name", name.ascii().data(), nullptr));
     gst_object_ref_sink(agent);
-    webkitGstWebRTCIceAgentConfigure(agent, WTFMove(socketProvider), context->identifier());
+    if (!webkitGstWebRTCIceAgentConfigure(agent, WTFMove(socketProvider), context->identifier())) {
+        gst_object_unref(agent);
+        return nullptr;
+    }
     return agent;
```
I would create a GRefPtr here and leak it at the return.

philn replied:
How would that integrate with the required `gst_object_ref_sink()` call?

calvaris replied:
Can't we have an specialization for `GRefPtr<GstObject>` and do something like:

```
    GRefPtr<GstObject> agent = reinterpret_cast<WebKitGstIceAgent*>(g_object_new(WEBKIT_TYPE_GST_WEBRTC_ICE_BACKEND, "name", name.ascii().data(), nullptr));
```

Then you can drop the unref and leak where you have to return.

philn replied:
No, we can't use a GRefPtr for the object creation like this because it will increase the refcount, and leaking it at return time will actually create a real leak.

calvaris replied:
Creating a GRefPtr without adoption from a floating point is going to sink the refcount, not increase it. Am I missing anything?

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -64,13 +64,27 @@ using WebKitGstRiceStream = struct _WebKitGstRiceStream {
+        if (onCandidateNotify)
+            onCandidateNotify(onCandidateData);
+
+        if (recvSource)
+            g_source_destroy(recvSource.get());
```
Why do we have to do this? Isn't the struct member already refcounted?

philn replied:
Pretty much everywhere GSource is used in WebKit, we call this before clearing the GRefPtr.

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp:
```diff
@@ -309,19 +325,30 @@ static gchar* webkitGstWebRTCIceAgentGetTurnServer(GstWebRTCICE* ice)
-    auto stream = GST_WEBRTC_ICE_STREAM(webkitGstWebRTCCreateIceStream(backend, WTFMove(riceStream)));
-    backend->priv->streams.append(WTF::makeUnique<WebKitGstRiceStream>(streamId, GRefPtr(stream)));
-    return stream;
+    auto stream = adoptGRef(GST_WEBRTC_ICE_STREAM(webkitGstWebRTCCreateIceStream(backend, WTFMove(riceStream))));
+    backend->priv->streams.add(sessionId, WTF::makeUnique<WebKitGstRiceStream>(streamId, WTFMove(stream)));
+    auto item = backend->priv->streams.find(sessionId);
```
I understand that you need to create the unique_ptr and then move it because unique_ptr does not have copy constructor. Then you `find` the bare pointer. I wonder if there's a better solution. Can we maybe make it GRefPtr?

philn replied:
I doubt it, let's think about this after landing this PR.

calvaris commented on Source/WebKit/NetworkProcess/webrtc/rice/RiceBackend.cpp:
```diff
@@ -261,7 +261,9 @@ void RiceBackend::finalizeStream(unsigned streamId)
         return true;
     });
-    m_sockets.remove(streamId);
+    auto data = m_sockets.take(streamId);
+    if (data.source)
+        g_source_destroy(data.source.get());
```
As I asked above a related question, can't we use a smart ptr here?

philn replied:
It's already a GRefPtr.

philn committed:
[GTK] webrtc/datachannel/multiple-connections.html flaky crashes in RiceBackendProxy::create()
https://bugs.webkit.org/show_bug.cgi?id=303366

Reviewed by Xabier Rodriguez-Calvar.

Handle IPC issues in RiceBackendProxy::create() and propagate backend creation failure to the
IceAgent. This patch also fixes various leaks in the Rice backend, including GSource leaks.

Canonical link: https://commits.webkit.org/304159@main
